wire-gson-support/src/test/java/com/squareup/wire/GsonAllStructsTest.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (C) 2024 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.wire
+
+import com.google.gson.Gson
+import com.google.gson.GsonBuilder
+import com.squareup.wire.json.assertJsonEquals
+import org.junit.Test
+import squareup.proto3.AllStructs
+
+class GsonAllStructsTest {
+
+  @Test
+  fun kotlinFullObject() {
+    val value = AllStructs(
+      struct = mapOf("key1" to "value1"),
+      list = listOf("a1", "a2"),
+      rep_struct = emptyList(),
+      rep_list = emptyList(),
+      rep_value_a = emptyList(),
+      oneof_struct = mapOf("key2" to "value2"),
+      map_int32_struct = mapOf(1 to mapOf("key3" to "value3")),
+      map_int32_list = mapOf(2 to listOf("c1", "c2")),
+      map_int32_value_a = mapOf(1 to mapOf("key4" to "value4")),
+      map_int32_null_value = mapOf(1 to null),
+    )
+
+    val json =
+      """
+      |{
+      |  "list": ["a1", "a2"],
+      |  "mapInt32List": {"2":["c1", "c2"]},
+      |  "mapInt32NullValue": {"1":null},
+      |  "mapInt32Struct": {"1":{"key3":"value3"}},
+      |  "mapInt32ValueA": {"1":{"key4":"value4"}},
+      |  "oneofStruct": {"key2":"value2"},
+      |  "struct": {"key1":"value1"}
+      |}
+      """.trimMargin()
+    assertJsonEquals(gson.toJson(value), json)
```

Comment:
does this test fail without your change? Just verifying!

Reply:
The test wasn't properly testing out the constructor creation which I just fixed. To your question, yes it does. Here is the issue: 

```
java.util.NoSuchElementException: ArrayDeque is empty.
	at kotlin.collections.ArrayDeque.removeFirst(ArrayDeque.kt:145)
	at com.squareup.wire.KotlinConstructorBuilder.build(KotlinConstructorBuilder.kt:114)
	at com.squareup.wire.MessageTypeAdapter.read(MessageTypeAdapter.kt:77)
	at com.squareup.wire.MessageTypeAdapter.read(MessageTypeAdapter.kt:25)
	at com.google.gson.TypeAdapter$1.read(TypeAdapter.java:204)
```

^ This is due to the custom re-ordering we were doing pre this PR